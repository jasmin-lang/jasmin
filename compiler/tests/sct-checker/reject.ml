open Jasmin
open Common
open Sct_checker_forward

let path = "fail"

let load_and_check name =
  Format.printf "File %s:@." name;
  let ((_, fds) as p) = load_file (Filename.concat path name) in
  List.iter
    (fun fd ->
      if fd.Prog.f_cc <> Internal then
        let f = fd.Prog.f_name.fn_name in
        match ty_prog Arch.is_ct_sopn p [ f ] with
        | exception Utils.HiError e ->
            Format.printf "Failed as expected %s: %a@." f Utils.pp_hierror e
        | _ ->
            Format.eprintf "Did not fail: %s.@." f;
            assert false)
    fds

let () =
  let cases = Sys.readdir path in
  Array.sort String.compare cases;
  Array.iter load_and_check cases
