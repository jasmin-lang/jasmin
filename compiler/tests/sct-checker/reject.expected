File basic.jazz:
Failed as expected missing_update_in_loop: "fail/basic.jazz", line 91 (2) to line 93 (3):
                                           speculative constant type checker: msf is Exact {
                                             } it should be be at least Exact {
                                            msf }
Failed as expected missing_update_after_loop: "fail/basic.jazz", line 84 (18-21):
                                              speculative constant type checker: MSF is Trans
                                              (! (x >u ((64u) 0)))
Failed as expected outdated_msf: "fail/basic.jazz", line 74 (9-12):
                                 speculative constant type checker: the variable msf is not known to be a msf, only {
                                   } are
Failed as expected conditional_init_msf: "fail/basic.jazz", line 65 (9-12):
                                         speculative constant type checker: the variable msf is not known to be a msf, only {
                                           } are
Failed as expected magic_msf: "fail/basic.jazz", line 56 (2-10):
                              speculative constant type checker: assignment to MSF variable msf not allowed
Failed as expected xof_init: "fail/basic.jazz", line 51 (9-14):
                             speculative constant type checker: return type for state is #[ptr = public, val = secret] it should be less than #[ptr = public, val = transient]
Failed as expected missing_then: "fail/basic.jazz", line 37 (18-19):
                                 speculative constant type checker: the variable m is not known to be a msf, only {
                                   } are
Failed as expected missing_else: "fail/basic.jazz", line 23 (18-19):
                                 speculative constant type checker: the variable m is not known to be a msf, only {
                                   } are
Failed as expected after_branch: "fail/basic.jazz", line 10 (18-19):
                                 speculative constant type checker: the variable m is not known to be a msf, only {
                                   } are
Failed as expected leak_transient: "fail/basic.jazz", line 1 (42-50):
                                   speculative constant type checker: x has type #transient but should be at most #public
File bug_887.jazz:
Failed as expected test_venv: "fail/bug_887.jazz", line 13 (3-4):
                              speculative constant type checker: r has type #secret but should be at most #public
File corruption.jazz:
Failed as expected does_corrupt_memory: "fail/corruption.jazz", line 26 (12-13):
                                        speculative constant type checker: return type for y is #[ptr = public, val = transient] it should be less than #[ptr = public, val = public]
Failed as expected corrupts_memory: "fail/corruption.jazz", line 12 (12-13):
                                    speculative constant type checker: return type for y is #[ptr = public, val = transient] it should be less than #[ptr = public, val = public]
File functions.jazz:
Failed as expected call_kills_msf: "fail/functions.jazz", line 9 (2-10):
                                   speculative constant type checker: 
                                   this function call destroys MSFs and { m } are required.
                                   Trace:
                                     the function leak destroys MSFs at "fail/functions.jazz", line 1 (36-48)
File inline-if-modmsf.jazz:
Failed as expected inline_if_modmsf: "fail/inline-if-modmsf.jazz", line 18 (2-11):
                                     speculative constant type checker: 
                                     this function call destroys MSFs and { ms } are required.
                                     Trace:
                                       the function f destroys MSFs at "fail/inline-if-modmsf.jazz", line 6 (8) to line 8 (9)
File modmsf-trace.jazz:
Failed as expected modmsf_trace: "fail/modmsf-trace.jazz", line 17 (2-8):
                                 speculative constant type checker: this function call destroys MSFs and { ms } are required.
                                 Trace:
                                   the function f2 destroys MSFs at "fail/modmsf-trace.jazz", line 12 (19-25)
                                   the function f1 destroys MSFs at "fail/modmsf-trace.jazz", line 9 (19-31)
                                   the function kill_msf destroys MSFs at "fail/modmsf-trace.jazz", line 3 (4) to line 5 (5)
File movmsf.jazz:
Failed as expected overwrite2: "fail/movmsf.jazz", line 29 (4-5):
                               speculative constant type checker: x cannot become an MSF as the current status depends on it (
                               (x >u ((64u) 0)))
Failed as expected overwrite: "fail/movmsf.jazz", line 21 (4-5):
                              speculative constant type checker: x cannot become an MSF as the current status depends on it (
                              (x >u ((64u) 0)))
Failed as expected fail: "fail/movmsf.jazz", line 13 (29-32):
                         speculative constant type checker: MSF is not Trans
File speculative-stack-leak.jazz:
Failed as expected main: "fail/speculative-stack-leak.jazz", line 32 (2) to line 36 (3):
                         speculative constant type checker: (pub > ((32u) 0)) has type #transient but should be at most #public
File spill.jazz:
Failed as expected spill2: "fail/spill.jazz", line 12 (5-8):
                           speculative constant type checker: pub has type #transient but should be at most #public
